apps/bfd-pipeline/bfd-pipeline-shared-utils/src/main/java/gov/cms/bfd/pipeline/sharedutils/SamhsaUtil.java

```diff
@@ -0,0 +1,392 @@
+  }
+
+  /**
+   * Builds a TagDetails object, and adds it to a list of TagDetails.
+   *
+   * @param entry The SamhsaEntry that holds information about the SAMHSA code.
+   * @param table The table that the code belongs to
+   * @param column The column that the cod belongs to
+   * @param lineNum The line number that the code belongs to
+   * @param detailsList The TagDetails list to add to
+   * @param serviceDate The first date of service for the claim
+   * @param throughDate The service through date
+   */
+  private void buildDetails(
+      Optional<SamhsaEntry> entry,
+      String table,
+      String column,
+      Integer lineNum,
+      List<TagDetails> detailsList,
+      LocalDate serviceDate,
+      LocalDate throughDate) {
+    if (entry.isPresent()) {
+      try {
+        LocalDate startDate = LocalDate.parse(entry.get().getStartDate());
+        LocalDate endDate =
+            entry.get().getEndDate().equals("Active")
+                ? LocalDate.now()
+                : LocalDate.parse(entry.get().getEndDate());
+        // if the last update to the claim is before the start date of the SAMHSA code
+        // or the service date of the claim is after the end date of the code,
+        // do nothing.
+        if (throughDate.isBefore(startDate) || serviceDate.isAfter(endDate)) {
+          return;
+        }
+      } catch (DateTimeParseException ignore) {
+        // Parsing the date from the SamhsaEntry failed, so the tag should be created by default.
+      }
+      // Use the last part of the system path as the type
+      String type =
+          Arrays.stream(entry.get().getSystem().split("/"))
+              .reduce((first, second) -> second)
+              .orElse(Strings.EMPTY);
+      TagDetails detail =
+          TagDetails.builder().table(table).column(column).clm_line_num(lineNum).type(type).build();
+      detailsList.add(detail);
+    }
+  }
+
+  /**
+   * Constructs a list of TagDetail objects for a Fiss claim.
+   *
+   * @param fissClaim The claim to check.
+   * @return a list of TagDetail objects, one for each SAMHSA code found in the claim.
+   */
+  private Optional<List<TagDetails>> getPossibleFissSamhsaFields(RdaFissClaim fissClaim) {
+    List<TagDetails> entries = new ArrayList<>();
+    LocalDate serviceDate =
+        fissClaim.getStmtCovFromDate() == null
+            ? LocalDate.parse("1970-01-01")
+            : fissClaim.getStmtCovFromDate();
+    LocalDate throughDate =
+        fissClaim.getStmtCovToDate() == null ? LocalDate.now() : fissClaim.getStmtCovToDate();
+
+    buildDetails(
+        isSamhsaCode(Optional.ofNullable(fissClaim.getAdmitDiagCode())),
+        "fiss_claims",
+        "admit_diag_code",
+        null,
+        entries,
+        serviceDate,
+        throughDate);
+    for (RdaFissRevenueLine revenueLine : fissClaim.getRevenueLines()) {
+      buildDetails(
+          isSamhsaCode(Optional.ofNullable(revenueLine.getApcHcpcsApc())),
```

Comment:
Might be worth a comment here stating that this column should never contain SAMHSA data, but we're checking it in case someone misinterprets this field.